File: backend/app/services/overflow.py
"""B overflow logic (charter §4.4).

Cap_overflow = 1.2 × P_active
If B_principal strictly > Cap_overflow, excess transfers to C.
"""

from decimal import Decimal


def calculate_overflow(b_principal: int, p_active: int) -> dict[str, int]:
    """Calculate B account overflow.

    Args:
        b_principal: B principal balance in cents
        p_active: Current target price in cents (0 = no wish list, skip overflow)

    Returns:
        Dict with cap_overflow, overflow_amount, b_after (all in cents)
    """
    # No wish list → no overflow constraint
    if p_active <= 0:
        return {
            "cap_overflow": 0,
            "overflow_amount": 0,
            "b_after": b_principal,
        }

    cap_overflow = int(Decimal(p_active) * Decimal("1.2"))

    if b_principal > cap_overflow:
        overflow_amount = b_principal - cap_overflow
        b_after = cap_overflow
    else:
        overflow_amount = 0
        b_after = b_principal

    return {
        "cap_overflow": cap_overflow,
        "overflow_amount": overflow_amount,
        "b_after": b_after,
    }

File: cloudfunctions/_shared/overflow.js
'use strict';

/**
 * Calculate B account overflow to C
 * Cap = floor(1.2 * P_active)
 * If B_principal > Cap: overflow = B - Cap
 *
 * @param {BigInt} bPrincipal - B balance in cents
 * @param {BigInt} pActive - P_active in cents
 * @returns {{ capOverflow: BigInt, overflowAmount: BigInt, bAfter: BigInt }}
 */
function calculateOverflow(bPrincipal, pActive) {
  const b = BigInt(bPrincipal);
  const p = BigInt(pActive);

  // If no wish list (P_active = 0), no overflow
  if (p === 0n) {
    return { capOverflow: 0n, overflowAmount: 0n, bAfter: b };
  }

  const capOverflow = p * 12n / 10n;

  if (b > capOverflow) {
    const overflow = b - capOverflow;
    return { capOverflow, overflowAmount: overflow, bAfter: capOverflow };
  }

  return { capOverflow, overflowAmount: 0n, bAfter: b };
}

module.exports = { calculateOverflow };
